crates/shared/src/price_estimation/native/coingecko.rs

```diff
@@ -71,101 +75,182 @@ impl CoinGecko {
+
+        Ok(token_price_eth
+            .checked_div(denominator_price_eth)
+            .context("division by zero")?
+            .to_f64()
+            .context("failed to convert price back to f64")?)
     }
 }
 
-impl NativePriceEstimating for CoinGecko {
-    fn estimate_native_price(&self, token: Token) -> BoxFuture<'_, NativePriceEstimateResult> {
+impl NativePriceBatchFetching for CoinGecko {
+    fn fetch_native_prices(
+        &'_ self,
+        requested_tokens: HashSet<H160>,
+    ) -> BoxFuture<'_, Result<HashMap<H160, NativePriceEstimateResult>, PriceEstimationError>> {
         async move {
+            let mut tokens = requested_tokens.iter().collect::<Vec<_>>();
             match self.quote_token {
-                QuoteToken::Eth => self.send_request_price_in_eth(token).await,
+                QuoteToken::Eth => {
+                    let prices = self.bulk_fetch_denominated_in_eth(&tokens).await?;
+                    Ok(tokens
+                        .into_iter()
+                        .map(|token| {
+                            let result = prices
+                                .get(token)
+                                .cloned()
+                                .ok_or(PriceEstimationError::NoLiquidity);
+                            (*token, result)
+                        })
+                        .collect())
+                }
                 QuoteToken::Other(native_price_token) => {
-                    let token_eth = rust_decimal::Decimal::try_from(
-                        self.send_request_price_in_eth(token).await?,
-                    )
-                    .map_err(|e| {
-                        PriceEstimationError::EstimatorInternal(anyhow!(
-                            "failed to parse requested token in ETH to rust decimal: {e:?}"
-                        ))
-                    })?;
-                    let native_price_token_eth = rust_decimal::Decimal::try_from(
-                        self.send_request_price_in_eth(native_price_token).await?,
-                    )
-                    .map_err(|e| {
-                        PriceEstimationError::EstimatorInternal(anyhow!(
-                            "failed to parse native price token in ETH to rust decimal: {e:?}"
-                        ))
-                    })?;
-                    let token_in_native_price =
-                        token_eth.checked_div(native_price_token_eth).ok_or(
-                            PriceEstimationError::EstimatorInternal(anyhow!("division by zero")),
-                        )?;
-                    token_in_native_price
-                        .to_f64()
-                        .ok_or(PriceEstimationError::EstimatorInternal(anyhow!(
-                            "failed to parse result to f64"
-                        )))
+                    if !requested_tokens.contains(&native_price_token) {
+                        tokens.push(&native_price_token);
```

Comment:
Are we pushing denominator here and then again inside `bulk_fetch_denominated_in_token`?

Reply:
We do. Good catch!
Fixed now.